controller/api/destination/external-workload/endpoints_controller.go

```diff
@@ -0,0 +1,547 @@
+package externalworkload
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	ewv1alpha1 "github.com/linkerd/linkerd2/controller/gen/apis/externalworkload/v1alpha1"
+	"github.com/linkerd/linkerd2/controller/k8s"
+	logging "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	// Name of the lease resource the controller will use
+	leaseName = "linkerd-destination-endpoint-write"
+
+	// Duration of the lease
+	leaseDuration = 30 * time.Second
+
+	// Deadline for the leader to refresh its lease. Core controllers have a
+	// deadline of 10 seconds.
+	leaseRenewDeadline = 10 * time.Second
+
+	// Duration a leader elector should wait in between action re-tries.
+	// Core controllers have a value of 2 seconds.
+	leaseRetryPeriod = 2 * time.Second
+
+	// Max retries for a service to be reconciled
+	maxRetryBudget = 15
+)
+
+// EndpointsController reconciles service memberships for ExternalWorkload resources
+// by writing EndpointSlice objects for Services that select one or more
+// external endpoints.
+type EndpointsController struct {
+	k8sAPI   *k8s.API
+	log      *logging.Entry
+	queue    workqueue.RateLimitingInterface
+	stop     chan struct{}
+	isLeader atomic.Bool
+
+	lec leaderelection.LeaderElectionConfig
+	sync.RWMutex
+}
+
+func NewEndpointsController(k8sAPI *k8s.API, hostname, controllerNs string, stopCh chan struct{}) (*EndpointsController, error) {
+	// TODO: pass in a metrics provider to the queue config
+	ec := &EndpointsController{
+		k8sAPI: k8sAPI,
+		queue: workqueue.NewRateLimitingQueueWithConfig(workqueue.DefaultControllerRateLimiter(), workqueue.RateLimitingQueueConfig{
+			Name: "endpoints_controller_workqueue",
+		}),
+		stop: stopCh,
+		log: logging.WithFields(logging.Fields{
+			"component": "external-endpoints-controller",
+		}),
+	}
+
+	_, err := k8sAPI.Svc().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ec.updateService,
+		DeleteFunc: ec.updateService,
+		UpdateFunc: func(_, newObj interface{}) {
+			ec.updateService(newObj)
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = k8sAPI.ExtWorkload().Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ew, ok := obj.(*ewv1alpha1.ExternalWorkload)
+			if !ok {
+				ec.log.Errorf("couldn't get ExternalWorkload from object %#v", obj)
+				return
+			}
+
+			if len(ew.Spec.WorkloadIPs) == 0 {
```

Comment:
Hm.. now that I think about it. Here is something you might encounter. We are not relying on resyncs, right. So we rely on the initial Adds once the controller fires up in order to get up to speed. What if we have the following scenario.

1. We have a workload with an IP and a service that selects it.
2. All is good and smooth
3. The controller crashes
4. Someone changes the workload to have zero IPs
5. The controller fires up again
6. We gen an add for this workload and ignore it because it has zero IPs

Can we end up in a situation where we have stale endpoints then? Or will these be resolved because we will get ADD callbacks on the Service informer as well?

Reply:
That should be resolved by looking at the service event, yes.

1. Controller crashes.
2. Someone makes changes to the workload to have zero IPs
3. Controller fires up again, does a full re-sync
4. We reconcile the service, get all of its _valid_ workloads
5. Detect a diff between current workload IPs and what's in the endpointslice (will contain IP from changed workload)
6. Remove stale address.